src/core/geometry/qgsgeometryengine.h

```diff
@@ -32,33 +33,46 @@ class QgsAbstractGeometryV2;
-
-    virtual void geometryChanged() = 0;
-    virtual void prepareGeometry() = 0;
-
-    virtual QgsAbstractGeometryV2* intersection( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* difference( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* combine( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* combine( const QList< QgsAbstractGeometryV2* >&, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* symDifference( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* buffer( double distance, int segments, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* buffer( double distance, int segments, int endCapStyle, int joinStyle, double mitreLimit, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* simplify( double tolerance, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* interpolate( double distance, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* envelope( QString* errorMsg = nullptr ) const = 0;
-    virtual bool centroid( QgsPointV2& pt, QString* errorMsg = nullptr ) const = 0;
-    virtual bool pointOnSurface( QgsPointV2& pt, QString* errorMsg = nullptr ) const = 0;
-    virtual QgsAbstractGeometryV2* convexHull( QString* errorMsg = nullptr ) const = 0;
-    virtual double distance( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual bool intersects( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual bool touches( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual bool crosses( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual bool within( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual bool overlaps( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual bool contains( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
-    virtual bool disjoint( const QgsAbstractGeometryV2& geom, QString* errorMsg = nullptr ) const = 0;
+    QgsGeometryEngine( const QgsAbstractGeometryV2* geometry, int precision = 0 );
+    ~QgsGeometryEngine();
+
+    /**
+     * Will clear any cached geometry data.
+     *
+     * TODO QGIS3: remove this and use implicitly shared QgsGeometry instead.
+     */
+    void geometryChanged();
+
+    /**
+     * Prepare the geomtetry for efficient processing.
+     *
+     * TODO: What is the optimal strategy for this in regards to different backends?
+     * Should the caller have to deal with this kind of lowlevel functionality?
+     * Do we need prepareGeosGeometry() instead?
```

Comment:
My preference is to have a separate class e.g. ```QgsPreparedGeometry``` for the prepared geometry functionality... The main reason being that it has very specific use case and the developer needs to understand its mechanics to have any positive effect:
- it works only for binary predicates
- it makes sense only when the geometry will be tested multiple times to offset initial cost of building the index
- it does not make sense to prepare both sides of the predicate